test/e2e/specs/editor/blocks/image.spec.js

```diff
@@ -869,63 +929,182 @@ test.describe( 'Image - interactivity', () => {
 			const postId = await editor.publishPost();
 			await page.goto( `/?p=${ postId }` );
 
-			openLightboxButton = page.getByRole( 'button', {
-				name: 'Enlarge image',
-			} );
-			lightbox = page.getByRole( 'dialog' );
-			closeButton = lightbox.getByRole( 'button', {
+			const lightbox = page.locator( '.wp-lightbox-overlay' );
+			await expect( lightbox ).toBeHidden();
+			const image = lightbox.locator( 'img' );
+
+			await expect( image ).toHaveAttribute( 'src', '' );
+
+			await page.getByRole( 'button', { name: 'Enlarge image' } ).click();
+
+			await expect( image ).toHaveAttribute(
+				'src',
+				new RegExp( filename )
+			);
+
+			await expect( lightbox ).toBeVisible();
+
+			const closeButton = lightbox.getByRole( 'button', {
 				name: 'Close',
 			} );
-		} );
+			await closeButton.click();
 
-		test( 'should open and focus appropriately using enter key', async ( {
-			page,
-		} ) => {
-			// Open and close lightbox using the close button
-			await openLightboxButton.focus();
-			await page.keyboard.press( 'Enter' );
-			await expect( lightbox ).toBeVisible();
-			await expect( closeButton ).toBeFocused();
+			await expect( lightbox ).toBeHidden();
 		} );
 
-		test( 'should close and focus appropriately using enter key on close button', async ( {
+		test( 'lightbox should be overriden when link is configured for image', async ( {
+			editor,
 			page,
 		} ) => {
-			// Open and close lightbox using the close button
-			await openLightboxButton.focus();
-			await page.keyboard.press( 'Enter' );
-			await expect( lightbox ).toBeVisible();
-			await expect( closeButton ).toBeFocused();
-			await page.keyboard.press( 'Enter' );
-			await expect( lightbox ).toBeHidden();
-			await expect( openLightboxButton ).toBeFocused();
+			await page.getByRole( 'button', { name: 'Advanced' } ).click();
+			const behaviorSelect = page.getByRole( 'combobox', {
+				name: 'Behaviors',
+			} );
+			await behaviorSelect.selectOption( 'lightbox' );
+
+			//await imageBlock.click();
+			await page
+				.getByLabel( 'Block tools' )
+				.getByLabel( 'Insert link' )
+				.click();
+
+			const form = page.locator(
+				'.block-editor-url-popover__link-editor'
+			);
+
+			const url = 'https://wordpress.org';
+
+			await form.getByLabel( 'URL' ).fill( url );
+
+			await form.getByRole( 'button', { name: 'Apply' } ).click();
+			await expect( behaviorSelect ).toBeDisabled();
+
+			const postId = await editor.publishPost();
+			await page.goto( `/?p=${ postId }` );
+
+			expect(
+				await page.locator( '.wp-lightbox-overlay' ).count()
+			).toEqual( 0 );
```

Comment:
I see! Thanks for the clarification!

Hmm, I guess now the problem is why do we want to make this assertion? It seems less ideal to me to assert something in a way that end users can't interact with. If we want to test if clicking on the overlay will close the lightbox then we should just write that test instead. If what we really want to test is the visibility of the scrim element then maybe we should select that instead?

These are implementation details and might change, for example, when we decide to use the native `dialog` element instead. Then we would have to refactor the test even though users' interactions haven't changed.

Now that I think about it, why don't we use the native [`::backdrop`](https://developer.mozilla.org/en-US/docs/Web/CSS/::backdrop) pseudo element in the first place? It seems to be supported in all the browsers we support.

---

With all that said, I'm just stating my observations, and I don't think any of it should block this wonderful PR! I'm just curious about the decisions we made around this area because I'm less informed about it 😅.

Reply:
Hmm, I guess now the problem is why do we want to make this assertion?

Thanks for the thoughts!

To clarify, this test is meant to verify the following: When a lightbox is configured for an image, the lightbox markup should NOT appear in the frontend if a link is enabled for the image. The reason we want to test this is that we cannot have both the lightbox and a link work simultaneously — we need to pick one. And so, enabling a link should override the lightbox functionality.

The ideas you’ve suggested may work for other scenarios, but not for this case unfortunately. Would be happy to hear of any other potential approaches.

Now that I think about it, why don't we use the native ::backdrop pseudo element in the first place? It seems to be supported in all the browsers we support.

I tried using ::backdrop, but for this implementation, we need to set the scrim to be the site background color, so using ::backdrop would have involved the use of CSS variables, and that seemed less straightforward and harder to follow than just setting the background color on a scrim div.

Just let me know if I can clarify anything further 🙏

Reply:
Yep! I should really re-read the test title 😅 .

In this case, I think checking user interactions could still work? A simple solution would be to test if the enlarge button still exists.

```js
await expect( page.getByRole( 'button', { name: 'Enlarge image' } ) ).not.toBeInViewport();
// `toBeAttached()` will be better but we don't support that API now. 😞 
```

What I'm trying to do is always trying to target user-visible elements with accessible attributes. In this case, `.wp-lightbox-overlay` is not visible and an implementation detail (unless we document it somewhere though!), but the enlarge button is accessible to screen readers. We can make sure that every time we change the test is because of some user flow changes. LMK if that helps! 🙇 

> so using ::backdrop would have involved the use of CSS variables, and that seemed less straightforward and harder to follow than just setting the background color on a scrim div.

Hmm, I personally don't think it's a problem TBH 😅 . If the spec is suggesting us to do this then I don't see why we should avoid using CSS variables in this case. I think it's just a matter of choosing the variable name? Something like...

```html
<!-- in PHP... -->
<div class="wp-lightbox-overlay" style="--wp-lightbox-scrim-color: $background_color;"></div>

<!-- and in SCSS... -->
<style>
.wp-lightbox-overlay::backdrop {
  background-color: var(--wp-lightbox-scrim-color);
}
</style>
```

Reply:
>```
>await expect( page.getByRole( 'button', { name: 'Enlarge image' } ) ).not.toBeInViewport();
>// `toBeAttached()` will be better but we don't support that API now. 😞 
>```

@kevin940726 Thanks for this! Agreed that this is more readable — I made this change.

>Hmm, I personally don't think it's a problem TBH 😅 . If the spec is suggesting us to do this then I don't see why we should avoid using CSS variables in this case. I think it's just a matter of choosing the variable name? Something like... (not tested)

Thanks for taking a look at this; we might be able to implement that — however, another issue with using `::backdrop` is that I'm unable to animate it, at least in my testing.

Here is the lightbox using the `scrim` div. Note the background fading in smoothly:

https://github.com/WordPress/gutenberg/assets/5360536/6eaf2f55-eaf2-437d-9f78-bf8b57cf906c

Contrast this with the following `::backdrop` test:

https://github.com/WordPress/gutenberg/assets/5360536/41551b61-3cf6-4bcd-a219-62df3ba7a3b8

As far as I can tell then, using `::backdrop` is not an option.

With this in mind, we could potentially switch the `.wp-lightbox-overlay` div to be a `dialog` element for semantic clarity in the tests and the markup. Note that when I was first implementing the lightbox, I just followed the convention that already existed in the codebase to iterate quickly and get feedback as quickly as possible. 

How the lightbox currently works is modeled on the responsive navigation menu implementation, which does not use the `dialog` _element_ but toggles the `dialog` _role_ on a div based on whether the modal is open. I suppose now is the time for us to revisit whether that implementation was the best approach or not.

So as far as semantic improvements go, we have a few options:

1. Leave the implementation as is
2. Leave the the `dialog` role on the `.wp-lightbox-overlay` whether the div is open or not
3. Switch the `.wp-lightbox-overlay` div to be a `dialog` element

That being said, I'm not sure which of these three approaches is the best option.
- In case of Option 1, I'm not sure why the navigation modal was implemented the way it was to begin with; perhaps there's a reason we're not aware of.
- With Option 2, I feel like this approach makes sense because, even though it would have the `dialog` role at all times, it has `aria-hidden` and `aria-modal="false"` to prevent it from being erroneously read by screenreaders.
- With Option 3, that makes sense as per the book, but we're not using the `dialog` element's other features, like `show()`, `showModal()`, or `::backdrop`, so perhaps it makes more sense to use a div to signal that it's a completely custom implementation.

These are my thoughts. What do you think?

I'm about to start working on more tests for the [lightbox zoom animation](https://github.com/WordPress/gutenberg/pull/51357) in a separate PR, so will be able to make any revisions then.